habitat-baselines/habitat_baselines/ml/mapping/dense/depth_utils.py

```diff
@@ -0,0 +1,158 @@
+import itertools
+from argparse import Namespace
+
+import numpy as np
+import torch
+
+from . import rotation_utils as ru
+
+
+def get_camera_matrix(width, height, fov):
+    """Returns a camera matrix from image size and fov."""
+    xc = (width - 1.0) / 2.0
+    zc = (height - 1.0) / 2.0
+    f = (width / 2.0) / np.tan(np.deg2rad(fov / 2.0))
+    camera_matrix = {"xc": xc, "zc": zc, "f": f}
+    camera_matrix = Namespace(**camera_matrix)
+    return camera_matrix
+
+
+def get_point_cloud_from_z_t(Y_t, camera_matrix, device, scale=1):
+    """Projects the depth image Y into a 3D point cloud.
+    Inputs:
+        Y is ...xHxW
+        camera_matrix
+    Outputs:
+        X is positive going right
+        Y is positive into the image
+        Z is positive up in the image
+        XYZ is ...xHxWx3
+    """
+    grid_x, grid_z = torch.meshgrid(
+        torch.arange(Y_t.shape[-1], device=device),
+        torch.arange(Y_t.shape[-2] - 1, -1, -1, device=device),
+    )
+    grid_x = grid_x.transpose(1, 0)
+    grid_z = grid_z.transpose(1, 0)
+    grid_x = grid_x.unsqueeze(0).expand(Y_t.size())
+    grid_z = grid_z.unsqueeze(0).expand(Y_t.size())
+
+    X_t = (
+        (grid_x[:, ::scale, ::scale] - camera_matrix.xc)
+        * Y_t[:, ::scale, ::scale]
+        / camera_matrix.f
+    )
+    Z_t = (
+        (grid_z[:, ::scale, ::scale] - camera_matrix.zc)
+        * Y_t[:, ::scale, ::scale]
+        / camera_matrix.f
+    )
+
+    XYZ = torch.stack(
+        (X_t, Y_t[:, ::scale, ::scale], Z_t), dim=len(Y_t.size())
+    )
+
+    return XYZ
+
+
+def transform_camera_view_t(
+    XYZ, sensor_height, camera_elevation_degree, device
+):
+    """
+    Transforms the point cloud into geocentric frame to account for
+    camera elevation and angle
+    Input:
+        XYZ                     : ...x3
+        sensor_height           : height of the sensor
+        camera_elevation_degree : camera elevation to rectify.
+    Output:
+        XYZ : ...x3
+    """
+    R = ru.get_r_matrix(
+        [1.0, 0.0, 0.0], angle=np.deg2rad(camera_elevation_degree)
+    )
+    XYZ = torch.matmul(
+        XYZ.reshape(-1, 3),
+        torch.from_numpy(R).float().transpose(1, 0).to(device),
```

Comment:
```suggestion
        torch.from_numpy(R).transpose(1, 0).to(device=device, dtype=torch.torch.float32),
```
Do casting and deice transfer at once